### DISCUSSION TITLE
Associating multiple SSH keys to single SSH keys

### DISCUSSION THREAD

**Discussion by kshuta** [robherley/snips.sh#164] [Ideas]
This might defeat the purpose of being passwordless, but what do you think about the possibility of associating multiple SSH keys to a single account / SSH key? I find myself wanting to access snips from different machines, but I'm not sure if there's a way to do that at the moment.

**Comment by robherley** [comment]
should be totally possible! a new command like [`ssh-import-id`](https://launchpad.net/ssh-import-id) where you can add a new key for your current user

the database is already setup such that a user:keys is 1:N: https://github.com/robherley/snips.sh/blob/8e52e02c66e74479d5643d22a2c6490cd73a5b7f/internal/db/schema.hcl#L23-L53

(might need a new index if we need to fine all keys for a user ID though)

e.g. something like taking a list of keys from stdin and adding them to your current user id

```
curl https://github.com/robherley.keys | ssh :import-pubkeys@snips.sh
```

(need to think of a nice pattern for having more "commands")

we'll need some more tui to add/remove keys too

**Comment by mokrinsky** [comment]
I just bumped into similar issue while rotating my ssh keys.
Input:
I use gpg-agent and all of my ssh keys are provided with gpg. I rotated my auth gpg key, hence ssh key changed. When I tried logging in into snips, I've seen that all of my snippets are gone.
Also, I use snips in kubernetes with self-written helm chart so I cannot dynamically update my keys using ssh-copy-id or smth.

Solution I used:
Snips uses sqlite as a database backend. So I opened database and what I've seen is:
```
sqlite> .headers ON
sqlite> select * from users;
id|created_at|updated_at
1|2023-07-18 03:39:14.784065001+00:00|2023-07-18 03:39:14.784065177+00:00
2|2024-08-24 14:39:30.56052203+00:00|2024-08-24 14:39:30.560522224+00:00
sqlite> select * from public_keys;;
id|created_at|updated_at|fingerprint|type|user_id
1|2023-07-18 03:39:15.077009676+00:00|2023-07-18 03:39:15.077029709+00:00|SHA256:fingerprint|ssh-rsa|1
2|2024-08-24 14:39:30.590117152+00:00|2024-08-24 14:39:30.590117293+00:00|SHA256:fingerprint|ssh-ed25519|2
```
Hence, every ssh key is associated with different user. Let's change it...
```
sqlite> update public_keys set user_id='1' where id='2';
sqlite> select * from public_keys;
id|created_at|updated_at|fingerprint|type|user_id
1|2023-07-18 03:39:15.077009676+00:00|2023-07-18 03:39:15.077029709+00:00|SHA256:fingerprint|ssh-rsa|1
2|2024-08-24 14:39:30.590117152+00:00|2024-08-24 14:39:30.590117293+00:00|SHA256:fingerprint|ssh-ed25519|1
```
What I've got is - different ssh keys are attached to the same user. Worked like a charm, now I can login using any of the keys and I get all of my snippets.

**Comment by ImDevinC** [comment]
Any update on this feature? I have multiple devices, each with their own SSH keys, and would like to be able to see all my snips uploaded from all my different devices.

**Comment by kshuta** [comment]
Sorry I haven't had the time to implement this feature, and am not sure if
I will in the near future.

**Comment by Lockszmith-GH** [comment]
Simplest would be to have a shared private key just for snips, and have a .ssh config for your snips.sh instance that always serves the same private key.
More to it, you can generate several ssh keys and share them with different teams.
I'll try it next week (I'm completely new to this awesome tool), and I'll post here an example.
